File: src/main/java/com/heybys/oddments/base/domain/ValueObject.java
package com.heybys.oddments.base.domain;

import java.util.Arrays;

public abstract class ValueObject<T extends ValueObject<T>> {

    @Override
    @SuppressWarnings("unchecked")
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        }

        if (!other.getClass().equals(getClass())) {
            return false;
        }

        return equals((T) other);
    }

    public boolean equals(T other) {
        if (other == null) {
            return false;
        }

        return Arrays.equals(getEqualityFields(), other.getEqualityFields());
    }

    @Override
    public int hashCode() {
        int hash = 17;
        for (Object each : getEqualityFields()) {
            hash = hash * 31 + (each == null ? 0 : each.hashCode());
        }
        return hash;
    }

    @SuppressWarnings({"java:S3011", "java:S112"})
    protected Object[] getEqualityFields() {
        return Arrays.stream(getClass().getDeclaredFields())
                .map(field -> {
                    field.setAccessible(true);
                    try {
                        return field.get(this);
                    } catch (IllegalAccessException e) {
                        throw new RuntimeException(e);
                    }
                })
                .toArray();
    }
}

File: src/main/kotlin/com/heybys/oddments/base/domain/ValueObject.kt
package com.heybys.oddments.base.domain

import java.util.Arrays

abstract class ValueObject<T : ValueObject<T>> {

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other.javaClass != javaClass) return false

        @Suppress("UNCHECKED_CAST")
        return equals(other as T)
    }

    open fun equals(other: T?): Boolean {
        if (other == null) return false
        return Arrays.equals(equalityFields, other.equalityFields)
    }

    override fun hashCode(): Int {
        var hash = 17
        for (field in equalityFields) {
            hash = hash * 31 + (field?.hashCode() ?: 0)
        }
        return hash
    }

    @Suppress("UNCHECKED_CAST")
    protected open val equalityFields: Array<Any?>
        get() =
            javaClass.declaredFields
                .map { field ->
                    field.isAccessible = true
                    try {
                        field.get(this)
                    } catch (e: IllegalAccessException) {
                        throw RuntimeException(e)
                    }
                }
                .toTypedArray()
}
